File: Tests/InstantTranslatorTests/LicenseServiceTests.swift
import Foundation
import XCTest
@testable import InstantTranslator

@MainActor
final class LicenseServiceTests: XCTestCase {

    // MARK: - Sign-in flow

    func test_requestMagicCode_setsPendingEmail_whenSucceeds() async {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let store = InMemoryAccountStore()
        let client = MockAuthClient()
        let (defaults, suiteName) = makeUserDefaults()
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        let ok = await service.requestMagicCode(email: " User@Example.com ")

        XCTAssertTrue(ok)
        XCTAssertEqual(service.pendingMagicCodeEmail, "user@example.com")
        XCTAssertEqual(client.requestedEmails, ["user@example.com"])
    }

    func test_requestMagicCode_rejectsInvalidEmail() async {
        let store = InMemoryAccountStore()
        let client = MockAuthClient()
        let (defaults, suiteName) = makeUserDefaults()
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        let ok = await service.requestMagicCode(email: "not-an-email")

        XCTAssertFalse(ok)
        XCTAssertNil(service.pendingMagicCodeEmail)
        XCTAssertEqual(client.requestedEmails, [])
    }

    func test_verifyMagicCode_savesSessionAndChecksStatus_whenSucceeds() async {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let store = InMemoryAccountStore()
        let client = MockAuthClient()
        client.verifyResult = .success("token-xyz")
        client.statusResult = .success(SubscriptionStatus(
            status: .active,
            planName: "Pro",
            trialDaysRemaining: nil
        ))
        let (defaults, suiteName) = makeUserDefaults()
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        _ = await service.requestMagicCode(email: "user@example.com")
        let ok = await service.verifyMagicCode("123456")

        XCTAssertTrue(ok)
        XCTAssertNil(service.pendingMagicCodeEmail)
        let saved = store.read()
        XCTAssertEqual(saved?.sessionToken, "token-xyz")
        XCTAssertEqual(saved?.email, "user@example.com")

        guard case .active(let planName, _, _) = service.state else {
            return XCTFail("Expected active state")
        }
        XCTAssertEqual(planName, "Pro")
    }

    func test_verifyMagicCode_keepsTrialState_whenSubscriptionInactive() async {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let store = InMemoryAccountStore()
        let client = MockAuthClient()
        client.verifyResult = .success("token-xyz")
        client.statusResult = .success(SubscriptionStatus(
            status: .trial,
            planName: nil,
            trialDaysRemaining: 7
        ))
        let (defaults, suiteName) = makeUserDefaults(installDate: now)
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        _ = await service.requestMagicCode(email: "user@example.com")
        _ = await service.verifyMagicCode("123456")

        if case .trial = service.state {
            // expected
        } else {
            XCTFail("Expected trial state, got \(service.state)")
        }
    }

    // MARK: - canTranslate gating

    func test_canTranslate_isFalse_whenStateExpired() {
        XCTAssertFalse(LicenseState.expired.canTranslate)
        XCTAssertFalse(LicenseState.invalid(reason: "x").canTranslate)
        XCTAssertFalse(LicenseState.checking.canTranslate)
    }

    func test_canTranslate_isTrue_whenTrialOrActive() {
        XCTAssertTrue(LicenseState.trial(daysRemaining: 1).canTranslate)
        XCTAssertTrue(LicenseState.active(planName: "Pro", validatedAt: Date(), isOfflineGrace: false).canTranslate)
        XCTAssertTrue(LicenseState.active(planName: "Pro", validatedAt: Date(), isOfflineGrace: true).canTranslate)
    }

    // MARK: - Offline grace

    func test_loadState_usesOfflineGrace_whenLastValidationWithin30Days() {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let validatedAt = now.addingTimeInterval(-15 * 24 * 60 * 60) // 15 days ago
        let store = InMemoryAccountStore()
        store.save(StoredAccountRecord(
            email: "user@example.com",
            sessionToken: "token-xyz",
            subscriptionStatus: "active",
            planName: "Pro",
            lastValidatedAt: validatedAt
        ))

        let client = MockAuthClient()
        let (defaults, suiteName) = makeUserDefaults()
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        guard case .active(_, _, let isOfflineGrace) = service.state else {
            return XCTFail("Expected active state, got \(service.state)")
        }
        XCTAssertTrue(isOfflineGrace)
    }

    func test_loadState_fallsBackToTrial_whenLastValidationOver30Days() {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let validatedAt = now.addingTimeInterval(-31 * 24 * 60 * 60)
        let store = InMemoryAccountStore()
        store.save(StoredAccountRecord(
            email: "user@example.com",
            sessionToken: "token-xyz",
            subscriptionStatus: "active",
            planName: "Pro",
            lastValidatedAt: validatedAt
        ))

        let client = MockAuthClient()
        let (defaults, suiteName) = makeUserDefaults(installDate: now)
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        if case .trial = service.state {
            // expected
        } else {
            XCTFail("Expected trial fallback, got \(service.state)")
        }
    }

    // MARK: - Sign out

    func test_signOut_clearsStoreAndPendingState() async {
        let now = Date(timeIntervalSince1970: 1_700_000_000)
        let store = InMemoryAccountStore()
        store.save(StoredAccountRecord(
            email: "user@example.com",
            sessionToken: "token",
            subscriptionStatus: "active",
            planName: "Pro",
            lastValidatedAt: now
        ))

        let client = MockAuthClient()
        let (defaults, suiteName) = makeUserDefaults(installDate: now)
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            now: { now },
            openURLHandler: { _ in },
            startBackgroundTasks: false
        )

        service.signOut()

        XCTAssertNil(store.read())
        XCTAssertNil(service.pendingMagicCodeEmail)
        XCTAssertFalse(service.isSignedIn)
    }

    // MARK: - openUpgrade / openManageSubscription

    func test_openUpgrade_callsCheckoutAndOpensURL() async {
        var openedURL: URL?
        let store = InMemoryAccountStore()
        store.save(StoredAccountRecord(
            email: "u@x.com", sessionToken: "tok", subscriptionStatus: nil,
            planName: nil, lastValidatedAt: nil
        ))
        let client = MockAuthClient()
        client.checkoutURL = URL(string: "https://checkout.stripe.com/c/pay/test")!
        let (defaults, suiteName) = makeUserDefaults()
        defer { cleanup(defaults, suiteName: suiteName) }

        let service = LicenseService(
            client: client,
            store: store,
            userDefaults: defaults,
            openURLHandler: { url in openedURL = url },
            startBackgroundTasks: false
        )

        await service.openUpgrade()

        XCTAssertEqual(openedURL?.absoluteString, "https://checkout.stripe.com/c/pay/test")
    }
}

// MARK: - Helpers

@MainActor
private func makeUserDefaults(installDate: Date? = nil) -> (UserDefaults, String) {
    let suiteName = "tests.licenseservice.\(UUID().uuidString)"
    let defaults = UserDefaults(suiteName: suiteName)!
    if let installDate = installDate {
        defaults.set(installDate, forKey: "installDate")
    }
    return (defaults, suiteName)
}

@MainActor
private func cleanup(_ defaults: UserDefaults, suiteName: String) {
    defaults.removePersistentDomain(forName: suiteName)
}

// MARK: - Test doubles

@MainActor
final class InMemoryAccountStore: AccountStoring {
    private var record: StoredAccountRecord?

    func read() -> StoredAccountRecord? {
        record
    }

    func save(_ record: StoredAccountRecord) {
        self.record = record
    }

    func delete() {
        record = nil
    }
}

@MainActor
final class MockAuthClient: MagicCodeAuthClientProtocol {
    enum MockError: Error { case notConfigured }

    var requestedEmails: [String] = []
    var verifyResult: Result<String, Error> = .success("default-token")
    var statusResult: Result<SubscriptionStatus, Error> = .success(
        SubscriptionStatus(status: .trial, planName: nil, trialDaysRemaining: 14)
    )
    var checkoutURL: URL = URL(string: "https://example.com/checkout")!
    var portalURL: URL = URL(string: "https://example.com/portal")!

    func requestMagicCode(email: String) async throws {
        requestedEmails.append(email)
    }

    func verifyMagicCode(email: String, code: String) async throws -> String {
        switch verifyResult {
        case .success(let token): return token
        case .failure(let error): throw error
        }
    }

    func refreshSubscriptionStatus(token: String) async throws -> SubscriptionStatus {
        switch statusResult {
        case .success(let status): return status
        case .failure(let error): throw error
        }
    }

    func createCheckoutSession(token: String) async throws -> URL {
        checkoutURL
    }

    func createPortalSession(token: String) async throws -> URL {
        portalURL
    }
}

File: windows/ControlV.Core.Tests/LicenseServiceTests.cs
using ControlV.Core;
using ControlV.Core.Models;
using Xunit;

namespace ControlV.Core.Tests;

public class LicenseServiceTests
{
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_700_000_000);

    private static LicenseService Make(MockAuthClient client, IAccountStore? store = null, IKeyValueStore? kv = null, DateTimeOffset? now = null)
        => new(client, store ?? new InMemoryAccountStore(), kv ?? new InMemoryKeyValueStore(), _ => { }, () => now ?? Now);

    [Fact]
    public async Task RequestMagicCode_NormalizesEmail_AndPersistsLastSignInEmail()
    {
        var client = new MockAuthClient();
        var kv = new InMemoryKeyValueStore();
        var service = Make(client, kv: kv);

        var ok = await service.RequestMagicCodeAsync(" User@Example.com ");

        Assert.True(ok);
        Assert.Equal("user@example.com", service.PendingMagicCodeEmail);
        Assert.Equal(new[] { "user@example.com" }, client.RequestedEmails);
        Assert.Equal("user@example.com", Make(new MockAuthClient(), kv: kv).LastSignInEmail);
    }

    [Fact]
    public async Task RequestMagicCode_RejectsInvalidEmail()
    {
        var client = new MockAuthClient();
        var service = Make(client);
        Assert.False(await service.RequestMagicCodeAsync("not-an-email"));
        Assert.Null(service.PendingMagicCodeEmail);
        Assert.Empty(client.RequestedEmails);
    }

    [Fact]
    public async Task VerifyMagicCode_SavesSession_AndBecomesActive()
    {
        var client = new MockAuthClient();
        var store = new InMemoryAccountStore();
        var service = Make(client, store);

        await service.RequestMagicCodeAsync("user@example.com");
        var ok = await service.VerifyMagicCodeAsync("123456");

        Assert.True(ok);
        Assert.Null(service.PendingMagicCodeEmail);
        Assert.Equal("token-xyz", store.Read()!.SessionToken);
        var active = Assert.IsType<LicenseState.Active>(service.State);
        Assert.Equal("Pro", active.PlanName);
        Assert.False(active.IsOfflineGrace);
    }

    [Fact]
    public async Task VerifyMagicCode_StaysOnTrial_WhenServerSaysTrial()
    {
        var client = new MockAuthClient { StatusResult = () => new SubscriptionStatus(SubscriptionStatusValue.Trial, null, 7) };
        var service = Make(client);
        await service.RequestMagicCodeAsync("user@example.com");
        await service.VerifyMagicCodeAsync("123456");
        Assert.IsType<LicenseState.Trial>(service.State);
    }

    [Fact]
    public void LoadState_UsesOfflineGrace_WithinThirtyDays()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", PlanName = "Pro", LastValidatedAt = Now.AddDays(-15) });
        var service = Make(new MockAuthClient(), store);
        var active = Assert.IsType<LicenseState.Active>(service.State);
        Assert.True(active.IsOfflineGrace);
    }

    [Fact]
    public void LoadState_FallsBackToTrialOrExpired_AfterThirtyDays()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", PlanName = "Pro", LastValidatedAt = Now.AddDays(-31) });
        var service = Make(new MockAuthClient(), store);
        Assert.IsType<LicenseState.Trial>(service.State); // fresh install date → 14-day trial
    }

    [Fact]
    public async Task Refresh_On401_DeletesSession_AndReportsExpiredSession()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", LastValidatedAt = Now.AddDays(-2) });
        var client = new MockAuthClient { StatusError = AuthException.Server(401, "Invalid session") };
        var service = Make(client, store);

        await service.RefreshSubscriptionStatusAsync(forceNetwork: true);

        Assert.Null(store.Read());
        Assert.Equal("Session expired. Please sign in again.", service.LastError);
        Assert.False(service.State is LicenseState.Active);
    }

    [Fact]
    public async Task Refresh_OnNetworkError_KeepsActive_WithinGrace()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", PlanName = "Pro", LastValidatedAt = Now.AddDays(-3) });
        var client = new MockAuthClient { StatusError = new HttpRequestException("offline") };
        var service = Make(client, store);

        await service.RefreshSubscriptionStatusAsync(forceNetwork: true);

        var active = Assert.IsType<LicenseState.Active>(service.State);
        Assert.True(active.IsOfflineGrace);
    }

    [Fact]
    public async Task Refresh_SkipsNetwork_WhenRecentlyValidated()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", PlanName = "Pro", LastValidatedAt = Now.AddHours(-1) });
        var client = new MockAuthClient { StatusError = new HttpRequestException("must not be called") };
        var service = Make(client, store);

        await service.RefreshSubscriptionStatusAsync(forceNetwork: false);

        var active = Assert.IsType<LicenseState.Active>(service.State);
        Assert.False(active.IsOfflineGrace);
    }

    [Fact]
    public void Trial_Expires_After14Days()
    {
        var kv = new InMemoryKeyValueStore();
        kv.SetString("installDate", Now.AddDays(-14).ToString("O"));
        var service = Make(new MockAuthClient(), kv: kv);
        Assert.IsType<LicenseState.Expired>(service.State);
        Assert.False(service.State.CanTranslate);
    }

    [Fact]
    public void SignOut_ClearsStore_KeepsLastSignInEmail()
    {
        var store = new InMemoryAccountStore();
        store.Save(new StoredAccountRecord { Email = "u@x.com", SessionToken = "tok", SubscriptionStatus = "active", LastValidatedAt = Now });
        var kv = new InMemoryKeyValueStore();
        kv.SetString("lastSignInEmail", "u@x.com");
        var service = Make(new MockAuthClient(), store, kv);

        service.SignOut();

        Assert.Null(store.Read());
        Assert.False(service.IsSignedIn);
        Assert.Equal("u@x.com", service.LastSignInEmail);
    }
}
